### PR TITLE
[SYSTEMML-1238] Change linear regression test to compare the results with command-line invocation since the intent is only to test the APIs

### PR DESCRIPTION
**Pull request by niketanpansare**
@iyounus can you please review this PR ?

### CODE REVIEW

**Comment by niketanpansare** [comment]
@iyounus  @bertholdreinwald  @mboehm7 This PR and the below code shows that LinRegDS has much less accuracy than scikit-learn, which implies bug in algorithm, default hyperparameters or something else:

```python
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /__ / .__/\_,_/_/ /_/\_\   version 2.1.0
      /_/

Using Python version 2.7.12 (default, Jul  2 2016 17:42:40)
SparkSession available as 'spark'.
>>> import numpy as np
>>> from pyspark.context import SparkContext
>>> from pyspark.ml import Pipeline
from pyspark.ml.feature import HashingTF, Tokenizer
>>> from pyspark.ml.feature import HashingTF, Tokenizer
from pyspark.sql import SparkSession
from sklearn import datasets, metrics, neighbors
>>> from pyspark.sql import SparkSession
>>> from sklearn import datasets, metrics, neighbors
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import TfidfVectorizer

from systemml.mllearn import LinearRegression, LogisticRegression, NaiveBayes, SVM
diabetes = datasets.load_diabetes()
diabetes_X = diabetes.data[:, np.newaxis, 2]
diabetes_X_train = diabetes_X[:-20]
diabetes_X_test = diabetes_X[-20:]
diabetes_y_train = diabetes.target[:-20]
diabetes_y_test = diabetes.target[-20:]
sparkSession = SparkSession.builder.getOrCreate()
regr = LinearRegression(sparkSession, solver="direct-solve")
regr.fit(diabetes_X_train, diabetes_y_train)
mllearn_y_predicted = regr.predict(diabetes_X_test).ravel()>>> from sklearn.datasets import fetch_20newsgroups
>>> from sklearn.feature_extraction.text import TfidfVectorizer
>>>
>>> from systemml.mllearn import LinearRegression, LogisticRegression, NaiveBayes, SVM
>>> diabetes = datasets.load_diabetes()
>>> diabetes_X = diabetes.data[:, np.newaxis, 2]
>>> diabetes_X_train = diabetes_X[:-20]
>>> diabetes_X_test = diabetes_X[-20:]
>>> diabetes_y_train = diabetes.target[:-20]
>>> diabetes_y_test = diabetes.target[-20:]
>>> sparkSession = SparkSession.builder.getOrCreate()
>>> regr = LinearRegression(sparkSession, solver="direct-solve")
>>> regr.fit(diabetes_X_train, diabetes_y_train)

Welcome to Apache SystemML!

17/02/17 13:56:52 WARN RewriteRemovePersistentReadWrite: Non-registered persistent write of variable 'X' (line 87).
17/02/17 13:56:52 WARN RewriteRemovePersistentReadWrite: Non-registered persistent write of variable 'y' (line 88).
BEGIN LINEAR REGRESSION SCRIPT
Reading X and Y...
Calling the Direct Solver...
Computing the statistics...
AVG_TOT_Y,153.36255924170615
STDEV_TOT_Y,77.21853383600028
AVG_RES_Y,4.8020565933360324E-14
STDEV_RES_Y,67.06389890324985
DISPERSION,4497.566536105316
PLAIN_R2,0.24750834362605834
ADJUSTED_R2,0.24571669682516795
PLAIN_R2_NOBIAS,0.24750834362605834
ADJUSTED_R2_NOBIAS,0.24571669682516795
Writing the output matrix...
END LINEAR REGRESSION SCRIPT
lr
>>> mllearn_y_predicted = regr.predict(diabetes_X_test).ravel()
17/02/17 13:56:54 WARN Expression: WARNING: null -- line 149, column 4 -- Read input file does not exist on FS (local mode):
17/02/17 13:56:54 WARN Expression: Metadata file:  .mtd not provided
>>>
... import numpy as np
>>> from sklearn import datasets
>>> diabetes = datasets.load_diabetes()
diabetes_X = diabetes.data[:, np.newaxis, 2]
>>> diabetes_X = diabetes.data[:, np.newaxis, 2]
>>> diabetes_X_train = diabetes_X[:-20]
>>> diabetes_X_test = diabetes_X[-20:]
>>> diabetes_y_train = diabetes.target[:-20]
>>> diabetes_y_test = diabetes.target[-20:]
>>> def writeColVector(X, fileName):
...     fileName = os.path.join(os.getcwd(), fileName)
...     X.tofile(fileName, sep='\n')
...     metaDataFileContent = '{ "data_type": "matrix", "value_type": "double", "rows":' + str(len(X)) + ', "cols": 1, "nnz": -1, "format": "csv", "author": "systemml-tests", "created": "0000-00-00 00:00:00 PST" }'
...     with open(fileName+'.mtd', 'w') as text_file:
...             text_file.write(metaDataFileContent)
...
>>> writeColVector(diabetes_X_test, 'lingreg_X_test.csv')
>>> writeColVector(diabetes_X, 'lingreg_X.csv')
>>> writeColVector(diabetes.target, 'lingreg_y.csv')
>>> from systemml import MLContext, dmlFromResource
>>> ml = MLContext(sc)
ml.execute(script)
>>> script = dmlFromResource('/scripts/algorithms/LinearRegDS.dml').input('$X',os.path.join(os.getcwd(),'lingreg_X.csv')).input('$Y', os.path.join(os.getcwd(),'lingreg_y.csv')).input('$B', os.path.join(os.getcwd(),'lingreg_B.csv')).input('$fmt', 'csv').input('$icpt', '1').input('$tol', '0.000001').input('$reg','1')
>>> ml.execute(script)
commandline_y_predicted = genfromtxt(os.path.join(os.getcwd(),'lingreg_predicted.csv'), delimiter=',').ravel()
17/02/17 13:57:54 WARN RewriteRemovePersistentReadWrite: Non-registered persistent read of variable 'X' (line 82).
17/02/17 13:57:54 WARN RewriteRemovePersistentReadWrite: Non-registered persistent read of variable 'y' (line 83).
17/02/17 13:57:54 WARN RewriteRemovePersistentReadWrite: Non-registered persistent write of variable 'X' (line 85).
17/02/17 13:57:54 WARN RewriteRemovePersistentReadWrite: Non-registered persistent write of variable 'y' (line 86).
17/02/17 13:57:54 WARN RewriteRemovePersistentReadWrite: Non-registered persistent write of variable 'beta_out' (line 226).
BEGIN LINEAR REGRESSION SCRIPT
Reading X and Y...
Calling the Direct Solver...
Computing the statistics...
AVG_TOT_Y,152.13348416289594
STDEV_TOT_Y,77.09300453299106
AVG_RES_Y,-2.935409582574532E-14
STDEV_RES_Y,66.48545020578437
DISPERSION,4420.315089065834
PLAIN_R2,0.2579428201690507
ADJUSTED_R2,0.2562563265785258
PLAIN_R2_NOBIAS,0.2579428201690507
ADJUSTED_R2_NOBIAS,0.2562563265785258
Writing the output matrix...
END LINEAR REGRESSION SCRIPT
MLResults
>>> script = dmlFromResource('/scripts/algorithms/GLM-predict.dml').input('$X',os.path.join(os.getcwd(),'lingreg_X_test.csv')).input('$M', os.path.join(os.getcwd(),'lingreg_predicted.csv')).input('$B', os.path.join(os.getcwd(),'lingreg_B.csv')).input('$fmt', 'csv').input('$icpt', '1').input('$tol', '0.000001').input('$reg','1')
>>> ml.execute(script)
17/02/17 13:57:55 WARN Expression: WARNING: null -- line 147, column 4 -- Read input file does not exist on FS (local mode):
17/02/17 13:57:55 WARN Expression: Metadata file:  .mtd not provided
17/02/17 13:57:55 WARN RewriteRemovePersistentReadWrite: Non-registered persistent read of variable 'X' (line 116).
17/02/17 13:57:55 WARN RewriteRemovePersistentReadWrite: Non-registered persistent read of variable 'B_full' (line 120).
17/02/17 13:57:55 WARN RewriteRemovePersistentReadWrite: Non-registered persistent write of variable 'means' (line 142).
MLResults
>>> from numpy import genfromtxt
>>> commandline_y_predicted = genfromtxt(os.path.join(os.getcwd(),'lingreg_predicted.csv'), delimiter=',').ravel()
>>> commandline_y_predicted
array([ 189.09660702,  133.32606012,  157.37391062,  132.81440373,
        135.88434209,  154.81562865,  194.21317095,  136.39599848,
        125.13955783,  137.41931127,  178.35182275,  123.60458865,
        152.76900308,  141.00090603,  116.95305553,  161.46716177,
        144.58250078,  144.58250078,  170.67697685,  117.46471193])
>>> mllearn_y_predicted
array([ 188.84521284,  134.98127765,  158.20701117,  134.4871131 ,
        137.45210036,  155.73618846,  193.78685827,  137.94626491,
        127.07464496,  138.93459399,  178.46775744,  125.59215133,
        153.75953028,  142.39374579,  119.16801227,  162.16032752,
        145.8528976 ,  145.8528976 ,  171.05528929,  119.66217681])
>>> diabetes_y_test
array([ 233.,   91.,  111.,  152.,  120.,   67.,  310.,   94.,  183.,
         66.,  173.,   72.,   49.,   64.,   48.,  178.,  104.,  132.,
        220.,   57.])
>>> from sklearn.metrics import accuracy_score, r2_score
>>> r2_score(diabetes_y_test, mllearn_y_predicted, multioutput='variance_weighted')
0.25395357789376738
>>> r2_score(diabetes_y_test, commandline_y_predicted, multioutput='variance_weighted')
0.27383106129322909
>>> from sklearn import datasets, linear_model
>>> sklearn_regr = linear_model.LinearRegression()
>>> sklearn_regr.fit(diabetes_X_train, diabetes_y_train)
LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)
>>> r2_score(diabetes_y_test, sklearn_regr.predict(diabetes_X_test), multioutput='variance_weighted')
0.47257544798227141
```

**Comment by niketanpansare** [comment]
It seems like regularization was the issue. After setting it to 0, we get same prediction as scikit-learn:
```python
>>> diabetes = datasets.load_diabetes()
>>> diabetes_X = diabetes.data[:, np.newaxis, 2]
>>> diabetes_X_train = diabetes_X[:-20]
>>> diabetes_X_test = diabetes_X[-20:]
>>> diabetes_y_train = diabetes.target[:-20]
>>> diabetes_y_test = diabetes.target[-20:]
>>> sparkSession = SparkSession.builder.getOrCreate()
>>> regr = LinearRegression(sparkSession, solver="direct-solve")
>>> regr.fit(diabetes_X_train, diabetes_y_train)

Welcome to Apache SystemML!

17/02/17 15:10:53 WARN RewriteRemovePersistentReadWrite: Non-registered persistent write of variable 'X' (line 87).
17/02/17 15:10:53 WARN RewriteRemovePersistentReadWrite: Non-registered persistent write of variable 'y' (line 88).
BEGIN LINEAR REGRESSION SCRIPT
Reading X and Y...
Calling the Direct Solver...
Computing the statistics...
AVG_TOT_Y,153.36255924170615
STDEV_TOT_Y,77.21853383600028
AVG_RES_Y,1.3705729547600037E-14
STDEV_RES_Y,63.03850633759284
DISPERSION,3973.853281274733
PLAIN_R2,0.33513125068675453
ADJUSTED_R2,0.3335482298550564
PLAIN_R2_NOBIAS,0.33513125068675453
ADJUSTED_R2_NOBIAS,0.3335482298550564
Writing the output matrix...
END LINEAR REGRESSION SCRIPT
lr
>>> mllearn_y_predicted = regr.predict(diabetes_X_test).ravel()
17/02/17 15:10:56 WARN Expression: WARNING: null -- line 149, column 4 -- Read input file does not exist on FS (local mode):
17/02/17 15:10:56 WARN Expression: Metadata file:  .mtd not provided
>>> mllearn_y_predicted
array([ 225.9732401 ,  115.74763374,  163.27610621,  114.73638965,
        120.80385422,  158.21988574,  236.08568105,  121.81509832,
         99.56772822,  123.83758651,  204.73711411,   96.53399594,
        154.17490936,  130.91629517,   83.3878227 ,  171.36605897,
        137.99500384,  137.99500384,  189.56845268,   84.3990668 ])
>>> from sklearn.metrics import accuracy_score, r2_score
>>> from sklearn import datasets, linear_model
>>> sklearn_regr = linear_model.LinearRegression()
>>> sklearn_regr.fit(diabetes_X_train, diabetes_y_train)
LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)
>>> sklearn_regr.predict(diabetes_X_test)
array([ 225.9732401 ,  115.74763374,  163.27610621,  114.73638965,
        120.80385422,  158.21988574,  236.08568105,  121.81509832,
         99.56772822,  123.83758651,  204.73711411,   96.53399594,
        154.17490936,  130.91629517,   83.3878227 ,  171.36605897,
        137.99500384,  137.99500384,  189.56845268,   84.3990668 ])

```

**Comment by iyounus** [comment]
LGTM
